--- linux-6.6.70/drivers/media/test-drivers/vivid/vivid-vid-common.c	2025-03-31 13:53:17.629453000 +0900
+++ linux-xlnx-6.6.70-v2024.1/drivers/media/test-drivers/vivid/vivid-vid-common.c	2025-04-05 13:10:32.916874600 +0900
@@ -74,6 +74,14 @@
 		.buffers = 1,
 	},
 	{
+		.fourcc   = V4L2_PIX_FMT_YUV444P,
+		.vdownsampling = { 1, 1, 1 },
+		.bit_depth = { 8, 8, 8 },
+		.color_enc = TGP_COLOR_ENC_YCBCR,
+		.planes   = 3,
+		.buffers = 1,
+	},
+	{
 		.fourcc   = V4L2_PIX_FMT_YUV420,
 		.vdownsampling = { 1, 2, 2 },
 		.bit_depth = { 8, 4, 4 },
