--- linux-6.6.70/drivers/video/hdmi.c	2025-03-31 13:53:24.921064100 +0900
+++ linux-xlnx-6.6.70-2024.1/drivers/video/hdmi.c	2025-03-31 13:55:16.767871600 +0900
@@ -60,7 +60,7 @@
 	memset(frame, 0, sizeof(*frame));
 
 	frame->type = HDMI_INFOFRAME_TYPE_AVI;
-	frame->version = 2;
+	frame->version = 3;
 	frame->length = HDMI_AVI_INFOFRAME_SIZE;
 }
 EXPORT_SYMBOL(hdmi_avi_infoframe_init);
@@ -68,7 +68,7 @@
 static int hdmi_avi_infoframe_check_only(const struct hdmi_avi_infoframe *frame)
 {
 	if (frame->type != HDMI_INFOFRAME_TYPE_AVI ||
-	    frame->version != 2 ||
+	    frame->version != 3 ||
 	    frame->length != HDMI_AVI_INFOFRAME_SIZE)
 		return -EINVAL;
 
@@ -160,7 +160,7 @@
 	if (frame->itc)
 		ptr[2] |= BIT(7);
 
-	ptr[3] = frame->video_code & 0x7f;
+	ptr[3] = frame->video_code & 0xff;
 
 	ptr[4] = ((frame->ycc_quantization_range & 0x3) << 6) |
 		 ((frame->content_type & 0x3) << 4) |
@@ -1635,7 +1635,7 @@
 	frame->quantization_range = (ptr[2] >> 2) & 0x3;
 	frame->nups = ptr[2] & 0x3;
 
-	frame->video_code = ptr[3] & 0x7f;
+	frame->video_code = ptr[3] & 0xff;
 	frame->ycc_quantization_range = (ptr[4] >> 6) & 0x3;
 	frame->content_type = (ptr[4] >> 4) & 0x3;
 
