diff --git a/drivers/rpmsg/Kconfig b/drivers/rpmsg/Kconfig
index d3795860f..5ed64ddf8 100644
--- a/drivers/rpmsg/Kconfig
+++ b/drivers/rpmsg/Kconfig
@@ -81,4 +81,16 @@ config RPMSG_VIRTIO
 	select RPMSG_NS
 	select VIRTIO
 
+config RPMSG_VIRTIO_BUF_SIZE
+	int
+	prompt "Virtio RPMSG buffer size"
+	default 512
+	depends on RPMSG_VIRTIO
+	help
+	  Maximum virtio buffer size in bytes for rx and tx virtqueues.
+	  Each buffer will use 16 bytes for the msg header and
+	  remaining for the payload.
+
+	  Leave as default if not sure.
+
 endmenu
diff --git a/drivers/rpmsg/virtio_rpmsg_bus.c b/drivers/rpmsg/virtio_rpmsg_bus.c
index f1af0f674..92de49d22 100644
--- a/drivers/rpmsg/virtio_rpmsg_bus.c
+++ b/drivers/rpmsg/virtio_rpmsg_bus.c
@@ -26,6 +26,7 @@
 #include <linux/virtio.h>
 #include <linux/virtio_ids.h>
 #include <linux/virtio_config.h>
+#include <uapi/linux/virtio_ring.h>
 #include <linux/wait.h>
 
 #include "rpmsg_internal.h"
@@ -110,15 +111,14 @@ struct virtio_rpmsg_channel {
 	container_of(_rpdev, struct virtio_rpmsg_channel, rpdev)
 
 /*
- * We're allocating buffers of 512 bytes each for communications. The
- * number of buffers will be computed from the number of buffers supported
- * by the vring, upto a maximum of 512 buffers (256 in each direction).
+ * We're allocating buffers of size 512 bytes by default if not configured
+ * by CONFIG_RPMSG_VIRTIO_BUF_SIZE. The number of buffers will be
+ * computed from the number of buffers supported by the vring,
+ * upto a maximum of 512 buffers (256 in each direction).
  *
- * Each buffer will have 16 bytes for the msg header and 496 bytes for
+ * Each buffer will have 16 bytes for the msg header and remaining for
  * the payload.
  *
- * This will utilize a maximum total space of 256KB for the buffers.
- *
  * We might also want to add support for user-provided buffers in time.
  * This will allow bigger buffer size flexibility, and can also be used
  * to achieve zero-copy messaging.
@@ -128,7 +128,12 @@ struct virtio_rpmsg_channel {
  * processor.
  */
 #define MAX_RPMSG_NUM_BUFS	(512)
-#define MAX_RPMSG_BUF_SIZE	(512)
+
+#ifndef CONFIG_RPMSG_VIRTIO_BUF_SIZE
+#define MAX_RPMSG_BUF_SIZE (512)
+#else
+#define MAX_RPMSG_BUF_SIZE CONFIG_RPMSG_VIRTIO_BUF_SIZE
+#endif
 
 /*
  * Local addresses are dynamically allocated on-demand.
@@ -557,6 +562,7 @@ static int rpmsg_send_offchannel_raw(struct rpmsg_device *rpdev,
 	struct virtio_rpmsg_channel *vch = to_virtio_rpmsg_channel(rpdev);
 	struct virtproc_info *vrp = vch->vrp;
 	struct device *dev = &rpdev->dev;
+	const struct vring *vr;
 	struct scatterlist sg;
 	struct rpmsg_hdr *msg;
 	int err;
@@ -620,6 +626,11 @@ static int rpmsg_send_offchannel_raw(struct rpmsg_device *rpdev,
 
 	dev_dbg(dev, "TX From 0x%x, To 0x%x, Len %d, Flags %d, Reserved %d\n",
 		src, dst, len, msg->flags, msg->reserved);
+
+	vr = virtqueue_get_vring(vrp->svq);
+	dev_dbg(dev, "TX vring: avail: %d used: %d\n", vr->avail->idx,
+		vr->used->idx);
+
 #if defined(CONFIG_DYNAMIC_DEBUG)
 	dynamic_hex_dump("rpmsg_virtio TX: ", DUMP_PREFIX_NONE, 16, 1,
 			 msg, sizeof(*msg) + len, true);
@@ -713,6 +724,7 @@ static int rpmsg_recv_single(struct virtproc_info *vrp, struct device *dev,
 	struct scatterlist sg;
 	bool little_endian = virtio_is_little_endian(vrp->vdev);
 	unsigned int msg_len = __rpmsg16_to_cpu(little_endian, msg->len);
+	const struct vring *vr;
 	int err;
 
 	dev_dbg(dev, "From: 0x%x, To: 0x%x, Len: %d, Flags: %d, Reserved: %d\n",
@@ -720,6 +732,11 @@ static int rpmsg_recv_single(struct virtproc_info *vrp, struct device *dev,
 		__rpmsg32_to_cpu(little_endian, msg->dst), msg_len,
 		__rpmsg16_to_cpu(little_endian, msg->flags),
 		__rpmsg32_to_cpu(little_endian, msg->reserved));
+
+	vr = virtqueue_get_vring(vrp->rvq);
+	dev_dbg(dev, "RX vring: avail: %d used: %d\n", vr->avail->idx,
+		vr->used->idx);
+
 #if defined(CONFIG_DYNAMIC_DEBUG)
 	dynamic_hex_dump("rpmsg_virtio RX: ", DUMP_PREFIX_NONE, 16, 1,
 			 msg, sizeof(*msg) + msg_len, true);
