diff --git a/drivers/media/platform/xilinx/xilinx-isppipeline.c b/drivers/media/platform/xilinx/xilinx-isppipeline.c
index 6c5555a8a..d358b0832 100644
--- a/drivers/media/platform/xilinx/xilinx-isppipeline.c
+++ b/drivers/media/platform/xilinx/xilinx-isppipeline.c
@@ -2413,7 +2413,7 @@ static int xisp_parse_of(struct xisp_dev *xisp)
  * it creates the first control from the array. If there are more controls
  * (size > 1), it iterates through the remaining controls and creates them.
  */
-static void xisp_create_controls(struct xisp_dev *xisp, u32 index,
+static inline void xisp_create_controls(struct xisp_dev *xisp, u32 index,
 				 struct v4l2_ctrl_config *ctrl_arr, u32 size)
 {
 	int itr;
